lib/MongoDB/_URI.pm

```diff
@@ -141,12 +156,166 @@ sub _parse_doc {
     return $set;
 }
 
+use Devel::Dwarn;
+sub _parse_options {
+    my ( $self, $valid, $result, $err_unsupported ) = @_;
+
+    my %parsed;
+    for my $opt ( split '&', $result->{options} ) {
+        my @kv = split '=', $opt, -1;
+        MongoDB::UsageError->throw("expected key value pair") unless @kv == 2;
+        my ( $k, $v ) = map { _unescape_all($_) } @kv;
+        # connection string spec calls for case normalization
+        ( my $lc_k = $k ) =~ tr[A-Z][a-z];
+        if ( !$valid->{$lc_k} ) {
+            if ( $err_unsupported ) {
+                MongoDB::Error->throw("Unsupported option '$k' in URI $self\n");
+            } else {
+                warn "Unsupported option '$k' in URI $self\n";
+            }
+            next;
+        }
+        if ( exists $parsed{$lc_k} && !exists $options_with_list_type{$lc_k} ) {
+            warn "Multiple options were found for the same value '$lc_k'\n";
```

Comment:
As long as I'm looking at this, we should have the warning message clarify that only the first one is used.